oxd-server/src/main/java/org/gluu/oxd/server/op/Validator.java

```diff
@@ -35,53 +37,107 @@
 
     private final OpenIdConfigurationResponse discoveryResponse;
     private AbstractJwsSigner jwsSigner;
-    private JwsSignerObject jwsSignerObject;
+    private final Jwt idToken;
+    private OpClientFactory opClientFactory;
+    private final PublicOpKeyService keyService;
+    private final Rp rp;
 
-    public Validator(JwsSignerObject jwsSignerObject, OpenIdConfigurationResponse discoveryResponse) {
-        Preconditions.checkNotNull(jwsSignerObject.getIdToken());
-        Preconditions.checkNotNull(discoveryResponse);
+    public OpenIdConfigurationResponse getDiscoveryResponse() {
+        return discoveryResponse;
+    }
+
+    public AbstractJwsSigner getJwsSigner() {
+        return jwsSigner;
+    }
+
+    public OpClientFactory getOpClientFactory() {
+        return opClientFactory;
+    }
+
+    public PublicOpKeyService getKeyService() {
+        return keyService;
+    }
 
-        this.discoveryResponse = discoveryResponse;
-        this.jwsSignerObject = jwsSignerObject;
-        this.jwsSigner = createJwsSigner(jwsSignerObject, discoveryResponse);
+    public Rp getRp() {
+        return rp;
+    }
+
+    private Validator(ValidatorBuilder builder) {
+        Preconditions.checkNotNull(builder.idToken);
+        Preconditions.checkNotNull(builder.discoveryResponse);
+
+        this.discoveryResponse = builder.discoveryResponse;
+        this.idToken = builder.idToken;
+        this.opClientFactory = builder.opClientFactory;
+        this.keyService = builder.keyService;
+        this.rp = builder.rp;
+        this.jwsSigner = createJwsSigner(idToken, discoveryResponse, keyService, opClientFactory, rp);
+    }
+
+    //Builder Class
+    public static class ValidatorBuilder {
+
+        // required parameters
+        private final OpenIdConfigurationResponse discoveryResponse;
+        private final Jwt idToken;
+        private OpClientFactory opClientFactory;
+        private final PublicOpKeyService keyService;
+        private final Rp rp;
+
+        public ValidatorBuilder(OpenIdConfigurationResponse discoveryResponse,
+                                Jwt idToken,
+                                OpClientFactory opClientFactory,
+                                PublicOpKeyService keyService,
+                                Rp rp) {
+
+            this.discoveryResponse = discoveryResponse;
+            this.idToken = idToken;
+            this.opClientFactory = opClientFactory;
+            this.keyService = keyService;
+            this.rp = rp;
+        }
+
+        public Validator build() {
```

Comment:
before building validator we should have preconditions. For example there is not sense in Validator if idToken is null.
`Preconditions.checkNotNull(idToken)`.

Reply:
Done